src/V3Timing.cpp

```diff
@@ -1062,58 +1044,102 @@ class TimingControlVisitor final : public VNVisitor {
+        if (netDelayp) {
+            if (nodep->timingControlp()) {
+                // If this assignment has a delay, create another one to handle the net delay
+                AstVarScope* const newvscp
+                    = createTemp(flp, m_contAsgnTmpNames.get(nodep), nodep->dtypep());
+                AstAssignW* assignp = new AstAssignW{
+                    nodep->fileline(), nodep->lhsp()->unlinkFrBack(),
+                    new AstVarRef{flp, newvscp, VAccess::READ}, netDelayp->cloneTree(false)};
+                assignp->user1(true);
+                nodep->addNextHere(assignp);
+                nodep->lhsp(new AstVarRef{flp, newvscp, VAccess::WRITE});
+            } else {
+                // Else just use this one with the net delay
+                nodep->timingControlp(netDelayp->cloneTree(false));
+            }
+        }
+        if (!nodep->timingControlp()) return;
+        // There will be some circular logic here, suppress the warning for newly created vars
         flp->modifyWarnOff(V3ErrorCode::UNOPTFLAT, true);
-        tempvscp->fileline(flp);
-        tempvscp->varp()->fileline(flp);
-        // Remap the LHS to the new temp var
-        nodep->lhsp(new AstVarRef{flp, tempvscp, VAccess::WRITE});
+        // Also suppress the warning for the LHS var, for cases like `assign #1 clk = ~clk;`
+        // TODO: Restore the warning for other, non-delayed drivers
+        nodep->lhsp()->foreach([](AstVarRef* refp) {
+            if (refp->access().isWriteOrRW()) {
+                refp->varp()->fileline()->modifyWarnOff(V3ErrorCode::UNOPTFLAT, true);
+            }
+        });
         // Convert it to an always; the new assign with intra delay will be handled by
         // visit(AstNodeAssign*)
         AstAlways* const alwaysp = nodep->convertToAlways();
-        visit(alwaysp);
-        // Put the LHS back in the AssignW; put the temp var on the RHS
-        nodep->lhsp(lhsp);
-        nodep->rhsp(new AstVarRef{flp, tempvscp, VAccess::READ});
-        // Put the AssignW right after the always. Different order can produce UNOPTFLAT on the LHS
-        // var
-        alwaysp->addNextHere(nodep);
+        visit(alwaysp);  // Visit now as we need to do some post-processing
+        VL_DO_DANGLING(nodep->deleteTree(), nodep);
+        // IEEE 1800-2023 10.3.3 - if the RHS value differs from the currently scheduled value to
+        // be assigned, the currently scheduled assignment is descheduled. To keep track if an
+        // assignment should be descheduled, each scheduled assignment event has a 'generation',
+        // and if at assignment time its generation differs from the current generation, it won't
+        // be performed
+        AstFork* const forkp = VN_AS(alwaysp->stmtsp(), Fork);
+        AstBegin* const beginp = VN_AS(forkp->stmtsp(), Begin);
+        AstAssign* const preAssignp = VN_AS(beginp->stmtsp(), Assign);
+        AstAssign* const postAssignp = VN_AS(preAssignp->nextp()->nextp(), Assign);
```

Comment:
How are these known to be non-null?  UASSERT_OBJ just to be safe with message saying why known non-null.